storage-proofs/porep/src/stacked/vanilla/cores.rs

```diff
@@ -0,0 +1,189 @@
+pub struct CoreIndex(usize);
+
+pub fn checkout_core_group() -> Option<MutexGuard<'static, CoreGroup>> {
+    match &*CORE_GROUPS {
+        Some(groups) => {
+            for (i, group) in groups.iter().enumerate() {
+                match group.try_lock() {
+                    Ok(guard) => {
+                        debug!("checked out core group {}", i);
+                        return Some(guard);
+                    }
+                    Err(_) => debug!("core group {} locked, could not checkout", i),
+                }
+            }
+            None
+        }
+        None => None,
+    }
+}
+
+pub fn bind_core(core_index: CoreIndex) -> Result<()> {
+    let child_topo = &TOPOLOGY.clone();
+    let tid = get_thread_id();
+    let mut locked_topo = child_topo.lock().unwrap();
+    let core = get_core_by_index(&locked_topo, core_index).map_err(|err| {
+        anyhow::format_err!("failed to get core at index {}: {:?}", core_index.0, err)
+    })?;
+
+    let cpuset = core.allowed_cpuset().ok_or_else(|| {
+        anyhow::format_err!("no allowed cpuset for core at index {}", core_index.0,)
+    })?;
+    debug!("allowed cpuset: {:?}", cpuset);
+    let mut bind_to = cpuset;
+
+    // Get only one logical processor (in case the core is SMT/hyper-threaded).
+    bind_to.singlify();
+
+    debug!("binding to {:?}", bind_to);
+    // Set the binding.
+    let result = locked_topo
+        .set_cpubind_for_thread(tid, bind_to, CPUBIND_THREAD)
+        .map_err(|err| anyhow::format_err!("failed to bind CPU: {:?}", err));
+
+    if result.is_err() {
+        warn!("error in bind_core, {:?}", result);
+    }
+
+    result
+}
+
+fn get_core_by_index<'a>(topo: &'a Topology, index: CoreIndex) -> Result<&'a TopologyObject> {
+    let idx = index.0;
+
+    match topo.objects_with_type(&ObjectType::Core) {
+        Ok(all_cores) if idx < all_cores.len() => Ok(all_cores[idx]),
+        Ok(all_cores) => Err(anyhow::format_err!(
+            "idx ({}) out of range for {} cores",
+            idx,
+            all_cores.len()
+        )),
+        _e => Err(anyhow::format_err!("failed to get core by index {}", idx,)),
+    }
+}
+
+/// Helper method to get the thread id through libc, with current rust stable (1.5.0) its not
+/// possible otherwise I think.
+#[cfg(not(target_os = "windows"))]
+fn get_thread_id() -> libc::pthread_t {
+    unsafe { libc::pthread_self() }
+}
+
+#[cfg(target_os = "windows")]
+fn get_thread_id() -> winapi::winnt::HANDLE {
+    unsafe { kernel32::GetCurrentThread() }
+}
+
+fn core_groups(cores_per_unit: usize) -> Option<Vec<Mutex<Vec<CoreIndex>>>> {
+    let topo = TOPOLOGY.lock().unwrap();
+
+    let core_depth = match topo.depth_or_below_for_type(&ObjectType::Core) {
+        Ok(depth) => depth,
+        Err(_) => return None,
+    };
+    let all_cores = topo.objects_with_type(&ObjectType::Core).unwrap();
+    let core_count = all_cores.len();
+
+    all_cores[0].allowed_cpuset().unwrap();
```

Comment:
I am not sure, but this might happen in the wild that `all_cores` is empty?

Reply:
Fair point. This would violate my expectation of how things work, but you're right that it's possible from a types/code perspective.

Reply:
Anyway, this was just a stray line to test callability of`allowed_cpuset()`, so will remove. Good eye.